lom/src/engine/context.go

```diff
@@ -0,0 +1,394 @@
+         * deregister of its actions.
+         */
+        cl.Close()
+        delete (p.activeClients, name)
+
+        for k, _ := range cl.Actions {
+            p.DeregisterAction(name, k)
+        }
+    }
+}
+
+func (p *ClientRegistrations_t) DeregisterAction(clName, actName string) {
+    if len(actName) == 0 {
+        LogError("Expect non empty name")
+    } else if r, ok := p.activeActions[actName]; ok {
+        if (len(clName) != 0) && (clName != r.Client) {
+            LogError("Skip de-register as action(%s) client(%s) != given(%s)",
+                    actName, r.Client, clName)
+        } else {
+            if cl, ok := p.activeClients[r.Client]; ok {
+                delete (cl.Actions, actName)
+            }
+            delete (p.activeActions, actName)
+            GetSeqHandler().DropRequest(actName)
+        }
+    }
+}
+
+func (p *ClientRegistrations_t) NotifyHeartbeats(actName string,
+            ts int64) {
+    if _, ok := p.activeActions[actName]; ok {
+        p.heartbeatCh <- actName
+    }
+}
+
+type HB_t struct {
+    Actions     []string
+    Timestamp   int64
+}
+
+type HBData_t struct {
+    LoM_Heartbeat HB_t
+}
+
+
+func (p *ClientRegistrations_t) PublishHeartbeats() {
+    /* Map will help combine duplicate heartbeats into one */
+    lst := make(map[string]struct{})
+
+    for {
+        /* Read inside the loop to help refresh any change */
+        hb := GetConfigMgr().GetGlobalCfgInt(ENGINE_HB_INTERVAL_SECS)
+        if hb == 0 {
+            hb = MIN_HB_INTERVAL_SECS
+        }
+        select {
+        case a := <- p.heartbeatCh:
+            lst[a] = struct{}{}
+            /* Collect actions */
+
+        case <- time.After(time.Duration(hb) * time.Second):
+            hb := &HBData_t { HB_t { make([]string, len(lst)), time.Now().Unix()}}
+
+            if len(lst) > 0 {
+                /* Publish collected actions, which could be empty */
+                i := 0
+                for k, _ := range lst {
+                    hb.LoM_Heartbeat.Actions[i] = k
+                    i++
+                }
+                /* Reset collected. */
+                lst = make(map[string]struct{})
+            }
+            PublishEvent(hb)
+        }
+    }
+}
+
+
+/* Request to be sent to client as response to client's recvServerRequest */
+func (p *ClientRegistrations_t) AddServerRequest(
+            actionName string, req *ServerRequestData) error {
+    if (len(actionName) == 0) || (req == nil) {
+        return LogError("Internal error: Nil args (%v) (%v)", actionName, req)
+    } else if a, ok := p.activeActions[actionName]; !ok {
+        return LogError("(%s): Action is not registered yet", actionName)
+    } else if cl, ok := p.activeClients[a.Client]; !ok {
+        return LogError("Internal error: client(%s) for action (%s) not found",
+                a.Client, actionName)
+    } else {
+        if len(cl.pendingWriteRequests) >= cap(cl.pendingWriteRequests) {
```

Comment:
Once this len & cap check are done, and channel is full, then immediately channel may have room again. How do we guarantee it ? 
if len(ch) == cap(ch) {
    // Channel was full, but might not be by now
} else {
    // Channel wasn't full, but might be by now
}
There is possibility that cl.pendingWriteRequests <- req may block leaving AddServerRequest() to block.

Reply:
Only one writer.